### PR TITLE
Update travis to use vulture during testing

### PR DESCRIPTION
**Pull request by alexander-yu**
Incorporating vulture into travis so that code can be linted for dead code fragments. Updated YAML files and scripts, and added a `whitelist.py` module, which is used by vulture as a list of functions/attributes to ignore.

Turns out that because vulture only performs static analysis, by design it's going to generate a number of false positives; it marks a lot of Flask routes as unused (unless they're being explicitly used elsewhere, it seems), as well as attribute assignments for things like bokeh figures (and probably assignment of attributes to pandas DataFrames, in a similar vein). In this sense the `whitelist.py` module is somewhat longish, but helpful to have.

vulture also ignores any unused local variables that start with an underscore.

### CODE REVIEW

File: density/db/db.py [hunk 1]
```diff
@@ -189,7 +189,7 @@ def get_oauth_code_for_uni(cursor, uni):
     else:
         # If the code DNE, create a new one and insert into the database.
         new_code = ''.join(random.choice(
-            string.ascii_uppercase + string.digits) for x in xrange(32))
+            string.ascii_uppercase + string.digits) for _x in xrange(32))
```
Comment (alanhdu) [inline density/db/db.py]: I personally prefer `__` (or `_`, but that sometimes conflicts with `_` as the last output in the repl) instead of `x`.

...We should also change this code. If it's for identification, we should use https://docs.python.org/2.7/library/uuid.html (either encoded as hex or as base64). If it's for security / crypto, we should rethink our lives and cry.

Related: do you mind adding a UNIQUE index + constraint onto the `oauth_data` schema? We might have to talk to @thebrianzeng about letting you run the build index command on the running postgres instance.

File: whitelist.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# Because vulture only performs static analysis, this file is used
+# by vulture as a whitelist module and passed as a parameter when
+# running vulture so that it'll know that certain methods are being
+# used.
+
```
Comment (alanhdu) [inline whitelist.py]: Is a `whitelist.py` the only way to do this, or can we somehow store it in `setup.cfg` or something?

Tangent: I'm so excited for https://www.python.org/dev/peps/pep-0518/ to become a thing so we can just put all our config in `pyproject.toml`

**Comment by alanhdu** [comment]
Also needs a rebase (get rid of the merge commits).

File: density/db/db.py (hunk 1 above)
Comment (alexander-yu) [inline density/db/db.py]: Yeah, I'll take a look into what this code is actually being used for. I'm not super familiar with postgres, but I can certainly try and take a look into adding those things into the schema.

File: whitelist.py (hunk 2 above)
Comment (alexander-yu) [inline whitelist.py]: While it doesn't explicitly have to be named `whitelist.py` or anything like that, from what it seems like, it just needs to be a Python module, which is unfortunate; this seems to be the method proposed by the developer. 

Looking through the vulture source code, it's not actually treating these modules any differently than other modules, and since vulture basically keeps track of functions/variables/etc. that have been used across the project, it looks like all this sort of whitelist module is doing is simply faking usage of variables/methods so that vulture will think that it's been used already and thus remove the errors. As a result, I don't we would be able to store it in some config file, which is unfortunate; the developer of vulture isn't supporting any special ignore/whitelisting feature, since I suppose this method does the job.

Yeah, the use of TOML is definitely gonna make config/package management a bit less complex; it's already enough work for Python.

File: whitelist.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+figure.yaxis.axis_line_color
+figure.yaxis.major_label_text_color
+figure.yaxis.major_label_orientation
+figure.yaxis.axis_line_width
+
+# Ignore Flask app attribute assignment
+density.app.json_encoder
+
+# Ignore all Flask routes
+density.get_connections
```
Comment (alanhdu) [inline whitelist.py]: Is there a way to express this more concisely? Maybe the `app` object contains a list of all its routes already?

File: whitelist.py (hunk 2 above)
Comment (alanhdu) [inline whitelist.py]: That's irritating. Looking at this, I'm not sure vulture is actually worthwhile anymore -- still, you've already done the work, so we might as well move forward on it.

File: density/db/db.py (hunk 1 above)
Comment (alanhdu) [inline density/db/db.py]: So the two options I see are:
- base64 encode `os.urandom` output (either concat it to a `uuid` to guarantee uniqueness or check for uniqueness when inserting to the database).
- Encrypt (probably via https://cryptography.io/en/latest/fernet/) a uuid.

I think the first one's a better choice, but your call.

Comment (alexander-yu) [inline density/db/db.py]: Yeah, I was thinking something along the lines of the first option, as well; I'll get on that, and see what I can eventually do about adding the UNIQUE index + constraint onto the `oauth_data` schema.

File: whitelist.py (hunk 3 above)
Comment (alexander-yu) [inline whitelist.py]: I've been experimenting, and definitely not; the static analysis performed by vulture imho isn't really extensive at all. It ignores scopes and only performs on object names; in particular, it performs analysis by building a syntax tree, but the problem is that you can spoof the names using random objects.

For example, `density.map` is unused according to vulture, but you can suppress this warning by simply declaring it in a whitelist module. But the `density` portion doesn't even need to exist; in fact, turns out it doesn't even need to be imported. Replacing the `density.map` declaration in the whitelist with something like `banana.map` similarly suppresses the warning.

To be honest, at this point I would actually be fine with not including vulture in the testing process; it seems like it's more trouble than it's worth.

Comment (alanhdu) [inline whitelist.py]: I'm with you -- I'm leaning towards not including vulture in the test suite. Your call though.

File: density/db/db.py (hunk 1 above)
Comment (alanhdu) [inline density/db/db.py]: Let's leave that for another PR.

**Comment by alanhdu** [comment]
If you want to merge the vulture stuff in, go ahead (LGTM), although I think it's probably more trouble than it's worth.

Do you want to obtain another PR about the `os.urandom` and API key nonsense?

**Comment by alexander-yu** [comment]
I'll just end up closing this PR; I'll make another PR for the API key stuff.
